### PR TITLE
Added a strategy for multipliers 

### PR DESCRIPTION
**Pull request by frederik1904**
Features:
Multiplier strategy, and an implementation that gives the original multiplier values
All the original types
An object for positions

All of it should be tested :)

### CODE REVIEW

File: src/main/kotlin/framework/Position.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package framework
+
+class Position(var row: Int, var col: Int) {
+    override fun equals(other: Any?): Boolean {
```
Comment (anbclausen) [inline src/main/kotlin/framework/Position.kt:4]: You should auto generate equals methods to ensure correctness. Also use bail out fast when possible.
Here's what I got from IntelliJ:
```
override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Position

        if (row != other.row) return false
        if (col != other.col) return false

        return true
    }
```

File: src/main/kotlin/framework/Position.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        }
+        return true
+    }
+
+    override fun hashCode(): Int {
+        return GameConstants.HashCodePrimeValue * row + col;
+    }
+
+    override fun toString(): String {
+        return String.format("[$row,$col]")
```
Comment (anbclausen) [inline src/main/kotlin/framework/Position.kt:20]: String.format is redundant. "[$row, $col]" works directly in Kotlin.

File: src/main/kotlin/framework/Position.kt (hunk 1 above)
Comment (anbclausen) [reply src/main/kotlin/framework/Position.kt:4]: This also means you do not have to develop the equals methods using TDD.

File: src/main/kotlin/framework/Types.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    ROCK("Rock"), STEEL("Steel"), WATER("Water");
+
+    override fun toString(): String {
+        return type
+    }
+}
+
+enum class EffectLevel(val multiplier: Double) {
+    Inefective(0.0), Weak(0.5), Normal(1.0), Strong(2.0);
+}
```
Comment (anbclausen) [inline src/main/kotlin/framework/Types.kt:16]: Consider moving these enums in GameConstants or give file another name implicating the responsibilities of the file. "Types" suggests that it only holds types.

**Review by anbclausen — CHANGES_REQUESTED** [review]
Also: 
1) OriginalMultiplierStrategy is really nice. Where did you get all the data?
2) We're currently discussing whether to use standard JUnit or Hamcrest. We'll let you know. :)

**Comment by frederik1904** [comment]
>     1. OriginalMultiplierStrategy is really nice. Where did you get all the data?
The data comes from this diagram: https://2.bp.blogspot.com/-H4BnLpEKS94/VF62mw1fm4I/AAAAAAAAAAk/oezpFub4UCw/s1600/types.jpg 😄

File: src/main/kotlin/framework/GameConstants.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    GROUND("Ground"), ICE("Ice"), NORMAL("Normal"), POISON("Poison"), PSYCHIC("Psychic"),
+    ROCK("Rock"), STEEL("Steel"), WATER("Water");
+
+    override fun toString(): String {
+        return type
+    }
+}
+
+enum class EffectLevel(val multiplier: Double) {
+    Inefective(0.0), Weak(0.5), Normal(1.0), Strong(2.0);
```
Comment (anbclausen) [inline src/main/kotlin/framework/GameConstants.kt:20]: Typo. Should be Ineffective.

File: src/test/kotlin/OriginalMultiplierStrategyTest.kt [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import framework.EffectLevel
+import framework.Types
+import junit.framework.Assert.assertTrue
```
Comment (anbclausen) [inline src/test/kotlin/OriginalMultiplierStrategyTest.kt:3]: We have settled on using Hamcrest. Can you please refactor these tests.

We're using 
import static org.hamcrest.CoreMatchers.*;
import static org.hamcrest.MatcherAssert.assertThat;

**Review by anbclausen — CHANGES_REQUESTED** [review]
Much better. I'll approve after fixing these two things :)

**Review by anbclausen — APPROVED** [review]
Looks good to me :)

**Comment by frederik1904** [comment]
@anbclausen mangler der ikke docs ?

File: src/main/kotlin/framework/GameConstants.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package framework
+
+object GameConstants {
+    val HashCodePrimeValue: Int = 337213
+
+}
+
+enum class Types(val type: String) {
```
Comment (soer7022) [inline src/main/kotlin/framework/GameConstants.kt:8]: Needs JavaDoc

File: src/main/kotlin/framework/GameConstants.kt [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    FIGHTING("Fighting"), FIRE("Fire"), FLYING("Flying"), GHOST("Ghost"), GRASS("Grass"),
+    GROUND("Ground"), ICE("Ice"), NORMAL("Normal"), POISON("Poison"), PSYCHIC("Psychic"),
+    ROCK("Rock"), STEEL("Steel"), WATER("Water");
+
+    override fun toString(): String {
+        return type
+    }
+}
+
+enum class EffectLevel(val multiplier: Double) {
```
Comment (soer7022) [inline src/main/kotlin/framework/GameConstants.kt:19]: Needs JavaDoc

File: src/main/kotlin/framework/Position.kt [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package framework
+
+class Position(var row: Int, var col: Int) {
```
Comment (soer7022) [inline src/main/kotlin/framework/Position.kt:3]: Needs JavaDoc

File: src/main/kotlin/framework/strategy/MultiplierStrategy.kt [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package framework.strategy
+import framework.EffectLevel
+import framework.Types
+
+interface MultiplierStrategy {
```
Comment (soer7022) [inline src/main/kotlin/framework/strategy/MultiplierStrategy.kt:5]: Needs JavaDoc

File: src/main/kotlin/standard/OriginalMultiplierStrategy.kt [hunk 10]
```diff
@@ -0,0 +1,77 @@
+package standard
+import framework.EffectLevel
+import framework.EffectLevel.*
+import framework.Types
+import framework.Types.*
+import framework.strategy.MultiplierStrategy
+
+class OriginalMultiplierStrategy : MultiplierStrategy {
+    private val effectivenessMap = mapOf(
```
Comment (soer7022) [inline src/main/kotlin/standard/OriginalMultiplierStrategy.kt:9]: This REALLY Needs JavaDoc

**Review by soer7022 — CHANGES_REQUESTED** [review]

**Review by soer7022 — APPROVED** [review]
Looks good to me

File: src/main/kotlin/standard/OriginalMultiplierStrategy.kt (hunk 10 above)
Comment (dangerstone) [inline src/main/kotlin/standard/OriginalMultiplierStrategy.kt:16]: neat! :D

File: src/main/kotlin/framework/Position.kt [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    override fun equals(other: Any?): Boolean {
+        if (other == null
```
Comment (dangerstone) [inline src/main/kotlin/framework/Position.kt]: Refactor to bail out fast

**Review by dangerstone — CHANGES_REQUESTED** [review]

**Comment by frederik1904** [comment]
@dangerstone tror du er kommet til at kigge på noget gammelt kode 😄

**Review by dangerstone — APPROVED** [review]

**Pull request merged**